interface/src/warmup_cooldown_allowance.rs

```diff
@@ -200,24 +196,46 @@ mod test {
     }
 
     #[test]
-    fn activation_overflow_path_returns_account_portion() {
-        // Force the u128 multiply to overflow: (u64::MAX * u64::MAX * 9) overflows u128.
-        // When that happens, the helper returns the full account_portion.
-        let current_epoch = 0;
-        let new_rate_activation_epoch = Some(0); // use "current" 9/100 to maximize multiplier
+    fn activation_overflow_scenario_still_rate_limits() {
+        // Extreme scenario where a single account holding nearly the total supply
+        // and tries to activate everything at once. Asserting rate limiting is maintained.
+        let supply_lamports: u64 = 400_000_000_000_000_000; // 400M SOL
+        let account_portion = supply_lamports;
         let prev = StakeHistoryEntry {
-            activating: 1,       // non-zero cluster_portion
-            effective: u64::MAX, // huge cluster_effective
+            activating: supply_lamports,
+            effective: supply_lamports,
             ..Default::default()
         };
-        let account_portion = u64::MAX;
+
         let result = calculate_activation_allowance(
-            current_epoch,
+            100,
             account_portion,
             &prev,
-            new_rate_activation_epoch,
+            None, // forces 25% rate
         );
-        assert_eq!(result, account_portion);
+
+        // Verify overflow actually occurs in this scenario
+        let rate_bps = ORIGINAL_WARMUP_COOLDOWN_RATE_BPS;
+        let would_overflow = (account_portion as u128)
+            .checked_mul(supply_lamports as u128)
+            .and_then(|n| n.checked_mul(rate_bps as u128))
+            .is_none();
+        assert!(would_overflow);
+
+        // The ideal result (with infinite precision) is 25% of the stake.
+        // 400M * 0.25 = 100M
+        let ideal_allowance = supply_lamports / 4;
+
+        // With saturation fix:
+        // Numerator saturates to u128::MAX (≈ 3.4e38)
+        // Denominator = 4e17 * 10,000 = 4e21
+        // Result = 3.4e38 / 4e21 ≈ 8.5e16 (85M SOL)
+        // 85M is ~21.25% of the stake (fail-safe)
+        // If we allowed unlocking the full account portion it would have been 100% (fail-open)
```

Comment:
Thanks for spelling this all out, it's really useful to see a concrete example of what could happen.

Can you add these as actual asserts in the test?